Adding of tool and gun mods and mind swapper in RND

hyperioo opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

* First PR related to RnD improvements (see RnD section of the KoT map).

* Tool and gun mods prites by INFRARED_BARON

* Mindswapper sprite by The Sprite Goblin

* To be merged after #5267 since it uses the new psy damage type

## Why It's Good For The Game

* Add high quality toolmods for late game and a few high-tech gunmods. Gunmods can be found in a new "Gunmods" category in the autolathe menu once researched.
* I hope the "battery_shunt" and "overdrive" gun mods will not be too much broken. In terms of numbers they reduce the total damage output of the equipped weapon but there is still a burst potential that could be a bit too strong.
* The mind swapping device can also be a fun thing to do at some point during the round (randomly swap the minds of all people near it).

## Changelog
:cl: Hyperio
add: Added new toolmods (engineering research node)
add: Added new gunmods (combat research node)
add: Added a mind swapping device (biotech research node)
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

SpadesSlick commented on code/game/machinery/mindswapper.dm:
```diff
@@ -0,0 +1,104 @@
+	idle_power_usage = 2
+	active_power_usage = 500
+
+/*/obj/machinery/mindswapper/New()
```
We don't keep commented code in our codebase, usually.

SpadesSlick commented on code/game/machinery/mindswapper.dm:
```diff
@@ -0,0 +1,104 @@
+/obj/machinery/mindswapper
+	name = "experimental mind swapper"
+	desc = "The name isn't descriptive enough?"
+	icon = 'icons/obj/kitchen.dmi'
```
It sure needs sprites.

TheShown911 commented on code/modules/projectiles/guns/mods/mods.dm:
```diff
@@ -200,4 +202,90 @@
+	GUN_UPGRADE_DAMAGE_RADIATION = 5)
+	I.prefix = "radiation-coating"
+	I.req_gun_tags = list(GUN_PROJECTILE)
+	I.gun_loc_tag = GUN_BARREL
```
No newline

SpadesSlick replied:
Please add new line.

SpadesSlick commented on code/modules/projectiles/guns/mods/mods.dm:
```diff
@@ -200,4 +202,90 @@
+/obj/item/weapon/gun_upgrade/mechanism/battery_shunt
+	name = "Moebius \"Thunder\" battery shunt"
+	desc = "This experimental battery shunt is a cutting edge tool attachment which bypasses battery protection circuits to deliver the maximum amount of power in the shortest amount of time."
+	icon_state = "dampener"
```
This one as well.

TheShown911 commented on code/modules/research/designs/circuits.dm:
```diff
@@ -423,4 +423,11 @@
+	name = "experimental mind swapper"
+	build_path = /obj/item/weapon/circuitboard/mindswapper
+	sort_string = "WAAAA"
+	category = CAT_MACHINE
```
No new line

SpadesSlick replied:
Same as above.

TheShown911 commented on code/modules/research/nodes/biotech.dm:
```diff
@@ -248,3 +248,20 @@
+	required_tech_levels = list()
+	cost = 3000
+
+	unlocks_designs = list(	/datum/design/research/circuit/mindswapper
+							)
```
No newline

AlexMorgan3817 replied:
`unlocks_designs = list(/datum/design/research/circuit/mindswapper)`

SpadesSlick commented on code/modules/projectiles/guns/mods/mods.dm:
```diff
@@ -200,4 +202,90 @@
+	name = "Moebius \"Thunder\" battery shunt"
+	desc = "This experimental battery shunt is a cutting edge tool attachment which bypasses battery protection circuits to deliver the maximum amount of power in the shortest amount of time."
+	icon_state = "dampener"
+	matter = list(MATERIAL_PLASTIC = 2, MATERIAL_PLASTEEL = 1, MATERIAL_PLATINUM = 1, MATERIAL_URANIUM = 1)
```
Platinum is a Bay snowflake material that's not used mostly anywhere. It would be too hard to print those.

GrayRachnid replied:
I actually like this since it means it needs miners to make it. Platinum is the only material that is rarely found in maintenance but miners get plenty of it.

hyperioo commented:
Thank you for your feedback @SpadesSlick and @TheShown911 

I added the new lines, replaced the platinum requirement by gold and remove the commented New() function of the mindswapper. I let the 
```
/*if (operating) // Overlay when the mindswapper is active
		src.overlays += image('icons/obj/kitchen.dmi', "gruse")*/
```
lines since when this machine will have its own sprite there will likely be an overlow to show that it's active.

SpadesSlick commented:
Also, dreamcheaker is failing.

hyperioo commented:
Replaced
```
var/list/swapBoddies = new
var/list/swapMinds = new
```
by
```
var/list/swapBoddies = list()
var/list/swapMinds = list()
```

hyperioo commented:
I'm looking into this, will push again.
```
code/game/machinery/mindswapper.dm, line 86, column 21:
warning: field access requires static type: "key"
- 72:3: add a type annotation after /list here
```

hyperioo commented:
It's ok now, it did not like the direct access to key with `(name_list[i]).key` so I've put a temporary variable.

AlexMorgan3817 commented on code/game/machinery/mindswapper.dm:
```diff
@@ -0,0 +1,97 @@
+		to_chat(user, SPAN_DANGER("The mind swapping process has been launched, there is no going back now."))
+		return
+	else
+		src.startswapping(user)
```
why the hell u use `src.` here?

AlexMorgan3817 commented on code/game/machinery/mindswapper.dm:
```diff
@@ -0,0 +1,97 @@
+	return 1
+
+/obj/machinery/mindswapper/proc/startswapping(mob/user as mob)
+	if(src.operating)
```
`.src`

AlexMorgan3817 commented on code/game/machinery/mindswapper.dm:
```diff
@@ -0,0 +1,97 @@
+
+	use_power(1000)
+	visible_message(SPAN_DANGER("You hear an increasingly loud humming coming from the mind swapper."))
+	src.operating = TRUE
```
`.src`

AlexMorgan3817 commented on code/game/machinery/mindswapper.dm:
```diff
@@ -0,0 +1,97 @@
+
+	spawn(swap_time)
+
+		src.operating = FALSE
```
`.src`

AlexMorgan3817 commented on code/game/machinery/mindswapper.dm:
```diff
@@ -0,0 +1,97 @@
+	if(stat & (NOPOWER|BROKEN))
+		return
+	/*if (operating) // Overlay when the mindswapper is active
+		src.overlays += image('icons/obj/kitchen.dmi', "gruse")*/
```
`.src` .-.

AlexMorgan3817 commented on code/game/objects/items/weapons/tools/mods/mod_types.dm:
```diff
@@ -390,6 +630,23 @@
+	desc = "A high-tech array of sharp bits of plasteel, seemingly adapted for easy affixing to a tool. Would make it into a better weapon, but won't do much for productivity."
+
+/obj/item/weapon/tool_upgrade/augment/spikes/hightech/New()
+	..()
```
. = ..()

AlexMorgan3817 commented on code/game/objects/items/weapons/tools/mods/mod_types.dm:
```diff
@@ -370,6 +595,21 @@
+	desc = "A high-tech adapter which more modifications to be attached to the tool.  A bit fragile but you can compensate."
+
+/obj/item/weapon/tool_upgrade/augment/expansion/hightech/New()
+	..()
```
. = ..()

AlexMorgan3817 commented on code/game/objects/items/weapons/tools/mods/mod_types.dm:
```diff
@@ -331,6 +542,20 @@
+	desc = "An auxiliary tank which stores 150 extra units of fuel at the cost of degradation."
+
+/obj/item/weapon/tool_upgrade/augment/fuel_tank/hightech/New()
+	..()
```
. = ..()

AlexMorgan3817 commented on code/game/objects/items/weapons/tools/mods/mod_types.dm:
```diff
@@ -406,7 +663,19 @@
+	desc = "A high-tech attachment that fits on almost everything, that gives a simple flat surface to employ the tool for hammering."
 
+/obj/item/weapon/tool_upgrade/augment/hammer_addon/hightech/New()
+	..()
```
. = ..()

TheShown911 replied:
..() > . = ..()

AlexMorgan3817 replied:
Ну это конечно да, но нет .-.

AlexMorgan3817 commented on code/modules/research/nodes/biotech.dm:
```diff
@@ -248,3 +248,20 @@
+	y = 0.6
+	icon = "medhud"
+
+	required_technologies = list(	/datum/technology/top_biotech
+								)
```
`	required_technologies = list(/datum/technology/top_biotech)`

AlexMorgan3817 commented on code/modules/research/nodes/combat.dm:
```diff
@@ -127,7 +127,27 @@
+	required_tech_levels = list()
+	cost = 4000
+
+	unlocks_designs = list(	/datum/design/research/item/weapon/gunmod/battery_shunt,
+							/datum/design/research/item/weapon/gunmod/overdrive,
+							/datum/design/research/item/weapon/gunmod/honker,
+							/datum/design/research/item/weapon/gunmod/toxin_coater,
+							/datum/design/research/item/weapon/gunmod/isotope_diffuser
 						)
```
```
	unlocks_designs = list(
							/datum/design/research/item/weapon/gunmod/battery_shunt,
							/datum/design/research/item/weapon/gunmod/overdrive,
							/datum/design/research/item/weapon/gunmod/honker,
							/datum/design/research/item/weapon/gunmod/toxin_coater,
							/datum/design/research/item/weapon/gunmod/isotope_diffuser
						)
```

AlexMorgan3817 commented on code/modules/research/nodes/combat.dm:
```diff
@@ -127,7 +127,27 @@
 
 	unlocks_designs = list(	/datum/design/research/item/weapon/decloner,
 							/datum/design/research/item/weapon/plasmapistol,
-							/datum/design/research/item/weapon_upgrade/penetrator
+							/datum/design/research/item/weapon/gunmod/penetrator
+						)
```
```
	unlocks_designs = list(
							/datum/design/research/item/weapon/decloner,
							/datum/design/research/item/weapon/plasmapistol,
							/datum/design/research/item/weapon_upgrade/penetrator
							/datum/design/research/item/weapon/gunmod/penetrator
						)
```

AlexMorgan3817 commented on code/game/machinery/mindswapper.dm:
```diff
@@ -83,7 +83,8 @@
 			ghost.mind.transfer_to(swapBoddies[i])
 			if(ghost.key)
-				(swapBoddies[i]).key = ghost.key	//have to transfer the key since the mind was not active
+				var/mob/living/L = swapBoddies[i]
+				L.key = ghost.key	//have to transfer the key since the mind was not active
```
```
				if(istype(L))
					L.key = ghost.key
```

AlexMorgan3817 commented:
And dude, can u screenshot what u have done with research trees?

hyperioo commented:
Will make the changes and the screenshots :)

hyperioo commented:
![](https://image.noelshack.com/fichiers/2020/31/3/1596050428-pr-rnd-part1.png)

hyperioo commented:
Should me good. Nodes circled in red are the new ones. 
Position of nodes ingineering and combat are unchanged. I shifted some nodes to the left in biotech to be able to read the description of the node without having to put the research popup in full screen.

AlexMorgan3817 commented on code/modules/research/nodes/engineering.dm:
```diff
@@ -235,6 +235,41 @@
+	required_tech_levels = list()
+	cost = 2500
+
+	unlocks_designs = list(	/datum/design/research/item/weapon/toolmod/stick_ht,
+							/datum/design/research/item/weapon/toolmod/heatsink_ht,
+							/datum/design/research/item/weapon/toolmod/plating_ht,
+							/datum/design/research/item/weapon/toolmod/guard_ht,
+							/datum/design/research/item/weapon/toolmod/ergonomicgrip_ht,
+							/datum/design/research/item/weapon/toolmod/ratchet_ht,
+							/datum/design/research/item/weapon/toolmod/redpaint_ht,
+							/datum/design/research/item/weapon/toolmod/whetstone_ht,
+							/datum/design/research/item/weapon/toolmod/dblade_ht,
+							/datum/design/research/item/weapon/toolmod/oxyjet_ht,
+							/datum/design/research/item/weapon/toolmod/motor_ht,
+							/datum/design/research/item/weapon/toolmod/laserguide_ht,
+							/datum/design/research/item/weapon/toolmod/stabilizedgrip_ht,
+							/datum/design/research/item/weapon/toolmod/magbit_ht,
+							/datum/design/research/item/weapon/toolmod/portedbarrel_ht,
+							/datum/design/research/item/weapon/toolmod/cellmount_ht,
+							/datum/design/research/item/weapon/toolmod/fueltank_ht,
+							/datum/design/research/item/weapon/toolmod/expansion_ht,
+							/datum/design/research/item/weapon/toolmod/spikes_ht,
+							/datum/design/research/item/weapon/toolmod/hammeraddon_ht
+							)
```
I think you should split it in moar research nodes

AlexMorgan3817 commented on code/modules/research/nodes/engineering.dm:
```diff
@@ -235,6 +235,41 @@
 
+/datum/technology/toolmods
+	name = "Toolmods"
+	desc = "High-tech toolmods."
```
Desc should hold more interesting description of designs or technology

AlexMorgan3817 commented on code/game/machinery/mindswapper.dm:
```diff
@@ -0,0 +1,98 @@
+	user.attack_log += "\[[time_stamp()]\] Triggered the mind swapper</b>"
+	msg_admin_attack("[user.name] ([user.ckey]) triggered the mind swapper (<A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[user.x];Y=[user.y];Z=[user.z]'>JMP</a>)")
+
+	spawn(swap_time)
```
Move spawn to timer
`addtimer(CALLBACK(src, .proc/foo), swap_time, TIMER_STOPPABLE))`

AlexMorgan3817 commented:
Rename RP in "Adding of tool and gun mods and mind swaper in RND"

hyperioo commented:
Splitted the toolmod node into several nodes, one for each category of toolmod.

Instead of just having upgraded versions of existing toolmods I created 8 new toolmods: plasmablock, rubbermesh, antistaining, booster, injector, compensatedbarrel, vibcompensator and hydraulic.

AlexMorgan3817 commented:
> Splitted the toolmod node into several nodes, one for each category of toolmod.
> 
> Instead of just having upgraded versions of existing toolmods I created 8 new toolmods: plasmablock, rubbermesh, antistaining, booster, injector, compensatedbarrel, vibcompensator and hydraulic.
> 

So show me now how <s>my girls</s> research trees looking! 👀

hyperioo commented:
@AlexMorgan3817 The 4 nodes are the wrenches on the right.

![](https://image.noelshack.com/fichiers/2020/32/1/1596473832-pr-rnd-part1-bis.png)

SpadesSlick commented:
You should use icons those of tool mods for nodes, same with gun mods.

hyperioo commented:
@AlexMorgan3817 Could you tell me where I should put the icons for the nodes ? Like there is a bluespace node which uses the "bluespacething" icon and I cannot find any reference to it.

AlexMorgan3817 commented:
> @AlexMorgan3817 Could you tell me where I should put the icons for the nodes ? Like there is a bluespace node which uses the "bluespacething" icon and I cannot find any reference to it.

nano/css/shared.css

hyperioo commented:
> 
> 
> > @AlexMorgan3817 Could you tell me where I should put the icons for the nodes ? Like there is a bluespace node which uses the "bluespacething" icon and I cannot find any reference to it.
> 
> nano/css/shared.css

Oh, so I need to update /nano/images/sciences.png and then set correctly the pixel positions ?

SpadesSlick commented:
It looks fine. Can we merge it?
